crates/uv-requirements/src/pyproject.rs

```diff
@@ -1,133 +1,490 @@
-    pub(crate) name: PackageName,
+    pub name: PackageName,
     /// Project dependencies
-    pub(crate) dependencies: Option<Vec<String>>,
+    pub dependencies: Option<Vec<String>>,
     /// Optional dependencies
-    pub(crate) optional_dependencies: Option<IndexMap<ExtraName, Vec<String>>>,
+    pub optional_dependencies: Option<IndexMap<ExtraName, Vec<String>>>,
     /// Specifies which fields listed by PEP 621 were intentionally unspecified
     /// so another tool can/will provide such metadata dynamically.
-    pub(crate) dynamic: Option<Vec<String>>,
+    pub dynamic: Option<Vec<String>>,
+}
+
+/// `tool`.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+pub struct Tool {
+    pub uv: Option<Uv>,
+}
+
+/// `tool.uv`.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+#[serde(deny_unknown_fields)]
+pub struct Uv {
+    pub sources: Option<HashMap<PackageName, Source>>,
+    pub workspace: Option<UvWorkspace>,
+}
+
+/// `tool.uv.workspace`.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+#[serde(deny_unknown_fields)]
+pub struct UvWorkspace {
+    pub members: Option<Vec<SerdePattern>>,
+    pub exclude: Option<Vec<SerdePattern>>,
+}
+
+/// (De)serialize globs as strings.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+pub struct SerdePattern(#[serde(with = "serde_from_and_to_string")] pub Pattern);
+
+#[cfg(feature = "schemars")]
+impl schemars::JsonSchema for SerdePattern {
+    fn schema_name() -> String {
+        <String as schemars::JsonSchema>::schema_name()
+    }
+
+    fn json_schema(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+        <String as schemars::JsonSchema>::json_schema(gen)
+    }
+}
+
+impl Deref for SerdePattern {
+    type Target = Pattern;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// A `tool.uv.sources` value.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+#[serde(untagged, deny_unknown_fields)]
+pub enum Source {
+    Git {
+        git: String,
+        subdirectory: Option<String>,
+        // Only one of the three may be used, we validate this later for a better error message.
+        rev: Option<String>,
+        tag: Option<String>,
+        branch: Option<String>,
+    },
+    Url {
+        url: String,
+        subdirectory: Option<String>,
+    },
+    Path {
+        path: String,
+        /// `false` by default.
+        editable: Option<bool>,
+    },
+    Registry {
+        // TODO(konstin): The string is more-or-less a placeholder
+        index: String,
+    },
+    Workspace {
+        workspace: bool,
+        /// `true` by default.
+        editable: Option<bool>,
+    },
```

Comment:
Can you add Rustdoc for each variant?

Reply:
I wrote those targeted at users since they are exported through json schema descriptions.